_episodes/16-parallel.md

```diff
@@ -25,632 +23,430 @@ We now have the tools we need to run a multi-processor job. This is a very
 
-Most of the calculations required to estimate π are in the
-`inside_circle` function:
+### MPI for Python
 
-1. Generating `n_samples` random values for `x` and `y`.
-1. Calculating `n_samples` values of `radii` from `x` and `y`.
-1. Counting how many values in `radii` are under 1.0.
+The Amdahl code has one dependency: __mpi4py__.
+If it hasn't already been installed on the cluster, `pip` will attempt to
+collect mpi4py from the Internet and install it for you.
+If this fails due to a one-way firewall, you must retrieve mpi4py on your
+local machine and upload it, just as we did for Amdahl.
 
-There's also one multiplication operation and one division operation required
-to convert the `counts` value to the final estimate of π in the main
-function.
+> ## Retrieve and Push mpi4py
+>
+> If installing Amdahl failed because mpi4py could not be installed,
+> retrieve the tarball from <https://github.com/mpi4py/mpi4py/tarball/master>
+> then `rsync` it to the cluster, extract, and install:
+>
+> ```
+> {{ site.local.prompt }} wget https://github.com/mpi4py/mpi4py/tarball/master -O mpi4py.tar.gz
+> {{ site.local.prompt }} scp amdahl.tar.gz {{ site.remote.user }}@{{ site.remote.login }}:
+> # or
+> {{ site.local.prompt }} rsync -avP amdahl.tar.gz {{ site.remote.user }}@{{ site.remote.login }}:
+> ```
+> {: .language-bash}
+> ```
+> {{ site.remote.prompt }} mkdir mpi4py
+> {{ site.remote.prompt }} tar -xvzf mpi4py.tar.gz -C amdahl --strip-components=1
+> {{ site.remote.prompt }} cd mpi4py
+> {{ site.remote.prompt }} python3 -m pip install --user .
+> {{ site.remote.prompt }} cd ../amdahl
+> {{ site.remote.prompt }} python3 -m pip install --user .
+> ```
+> {: .language-bash}
+{: .discussion}
 
-A simple way to measure the calculation time is to use Python's `datetime`
-module to store the computer's current date and time before and after the
-calculations, and calculate the difference between those times.
+`pip` may warn that your user package binaries are not in your PATH.
 
-To add the time measurement to the script, add the following line below the
-`import sys` line:
+> ## If Pip Raises a Warning...
+>
+> ```
+> WARNING: The script amdahl is installed in "${HOME}/.local/bin" which is
+> not on PATH. Consider adding this directory to PATH or, if you prefer to
+> suppress this warning, use --no-warn-script-location.
+> ```
+> {: .warning}
+>
+> To check whether this warning is a problem, use `which` to search for the
+> `amdahl` program:
+>
+> ```
+> {{ site.remote.prompt }} which amdahl
+> ```
+> {: .language-bash}
+>
+> If the command returns no output, displaying a new prompt, it means the file
+> `amdahl` has not been found. You must update the environment variable named
+> `PATH` to include the missing folder.
+> Run the following command to update your shell configuration file, then log
+> off the cluster and back on again so it takes effect.
+>
+> ```
+> {{ site.remote.prompt }} echo "export PATH=${PATH}:${HOME}/.local/bin" >> ~/.bashrc
```

Comment:
Getting this wrong is extremely dangerous and can break their login shell...I can't think of any way out though

Reply:
Explicitly adding the line via an editor may be the better option

Reply:
Two things:

1) Beware that some clusters don't use bash as the shell, so it will not work in those esoteric cluster. (Our cluster is one of those, unless the user is savvy enough to switch to bash.)

2) This command leverages two (rather) advanced knowledge as far as the basic Software Carpentry's `shell-novice` lesson is concerned: (a) shell variables; (b) customization using .bashrc. Both are actually covered in the `shell-extra` lesson:

  - https://carpentries-incubator.github.io/shell-extras/07-aliases/index.html
  - https://carpentries-incubator.github.io/shell-extras/08-environment-variables/index.html

Do we need to provide explanation or link for those that may be curious about the syntax, etc?
Learners will need to be warned that it MUST be two `>>` instead of single `>`. It can cause havoc on the user's environment. Another alternative is to edit `~/.bashrc` with nano, though other things could go wrong in this way.

Reply:
Editing with nano seems safer and portable.

Reply:
Agreed, 9b656ad modifies this to use `nano`.